pdns/ssqlite3.cc

```diff
@@ -234,19 +234,28 @@ std::unique_ptr<SSqlStatement> SSQLite3::prepare(const string& query, int nparam
 
 void SSQLite3::execute(const string& query) {
   char *errmsg;
-  int rc;
-  if (sqlite3_exec(m_pDB, query.c_str(), NULL, NULL, &errmsg) == SQLITE_BUSY) {
+  std::string errstr1;
+  int rc = sqlite3_exec(m_pDB, query.c_str(), NULL, NULL, &errmsg);
+  if (rc != 0) {
+    errstr1 = errmsg;
+    sqlite3_free(errmsg);
+  }
+  if (rc == SQLITE_BUSY) {
     if (m_in_transaction) {
-      std::string errstr(errmsg);
-      sqlite3_free(errmsg);
-      throw("Failed to execute query: " + errstr);
+      throw("Failed to execute query: " + errstr1);
     } else {
-      if ((rc = sqlite3_exec(m_pDB, query.c_str(), NULL, NULL, &errmsg) != SQLITE_OK) && rc != SQLITE_DONE && rc != SQLITE_ROW) {
-        std::string errstr(errmsg);
+      rc = sqlite3_exec(m_pDB, query.c_str(), NULL, NULL, &errmsg);
+      std::string errstr2;
+      if (rc != 0)  {
+        errstr2 = errmsg;
         sqlite3_free(errmsg);
-        throw("Failed to execute query: " + errstr);
+      }
+      if (rc != SQLITE_OK && rc != SQLITE_DONE && rc != SQLITE_ROW) {
```

Comment:
Looking at the documentation I don't think we can get `SQLITE_DONE` or `SQLITE_ROW` as a result of calling `sqlite3_exec()` anyway.